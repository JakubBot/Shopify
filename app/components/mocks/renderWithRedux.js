import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
const defaultInitialState = {
  user: {
    name: '',
    email: '',
    token: '',
  },
  products: [],
  shippingAddress: {
    fullName: '',
    address: '',
    city: '',
    postalCode: '',
  },
};

function defaultReducer(state = defaultInitialState, action) {
  switch (action.type) {
    default:
      return state;
  }
}

function renderWithRedux(
  component,
  { reducer = defaultReducer, initialState = defaultInitialState } = {}
) {
  const store = createStore(reducer, initialState);

  return render(<Provider store={store}>{component}</Provider>);
}

export default renderWithRedux;
