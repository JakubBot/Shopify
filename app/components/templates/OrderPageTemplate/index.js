import Header from '@element/Header';
import Footer from '@element/Footer';
import { connect } from 'react-redux';
import styles from './index.module.scss';
import { useEffect } from 'react';
import { useRouter } from 'next/router';
import uniqid from 'uniqid';
import { useState } from 'react';
import { totalPricesProducts } from '@util/helperFunctions';
import { PayPalScriptProvider, PayPalButtons } from '@paypal/react-paypal-js';
import axios from 'axios';
import * as productActions from '@redux/actions/productActions';
import { toast } from 'react-toastify';
import { getError } from '@util/error';

const OrderPageTemplate = ({ products, shippingAddress, user, clearCart }) => {
  const initialOptions = {
    'client-id': process.env.NEXT_PUBLIC_CLIENT_ID_PAYPAL,
    currency: 'USD',
  };

  const router = useRouter();
  const [productsPrice, setProductsPrice] = useState(0);
  useEffect(() => {
    if (!user.token) router?.push('/login?redirect=/order');
    if (Object.values(shippingAddress).some((adress) => adress === ''))
      router?.push('/shippingAddress');
  }, [shippingAddress, router, user]);

  useEffect(() => {
    let totalPrices = totalPricesProducts(products);
    setProductsPrice(totalPrices);
  }, [products]);

  const shipping = {
    fullName: {
      value: shippingAddress.fullName,
      title: 'Full Name',
    },
    address: {
      value: shippingAddress.address,
      title: 'Address',
    },
    city: {
      value: shippingAddress.city,
      title: 'City',
    },
    postalCode: {
      value: shippingAddress.postalCode,
      title: 'Postal Code',
    },
  };
  // function saveOrder() {
  //   try {
  //     axios.post(
  //       '/api/orders',
  //       { products, shippingAddress },
  //       {
  //         headers: {
  //           authorization: `Bearer ${user.token}`,
  //         },
  //       }
  //     );
  //     clearCart();
  //     toast.success('Order was saved', {
  //       theme: 'colored',
  //     });
  //     router.push('/');
  //   } catch (err) {
  //     alert(err);
  //   }
  // }
  function onApprove(data, actions) {
    return actions.order.capture().then(async function () {
      try {
        axios.post(
          '/api/orders',
          { products, shippingAddress },
          {
            headers: {
              authorization: `Bearer ${user.token}`,
            },
          }
        );
        clearCart();
        toast.success('Order was saved', {
          theme: 'colored',
        });
        router.push('/');
      } catch (err) {
        toast.error(getError(err), {
          theme: 'colored',
        });
      }
    });
  }

  return (
    <div className={styles.orderContainer}>
      <Header />
      <div className={styles.order}>
        <div className={styles.orderWrapper}>
          <div className={styles.orderProducts}>
            <h3 className={styles.orderTitle}>Your Products</h3>
            <ul className={styles.productList}>
              <li className={styles.product}>
                <span className={`${styles.productCell} ${styles.productName}`}>
                  Name
                </span>
                <span
                  className={`${styles.productCell} ${styles.productPrice}`}
                >
                  Price
                </span>
                <span
                  className={`${styles.productCell} ${styles.productQuantity}`}
                >
                  Quantity
                </span>
              </li>
              <ol aria-label="products" className={styles.productList}>
                {products.map(({ name, price, quantity, id }) => {
                  return (
                    <li key={id} className={styles.product}>
                      <span className={styles.productCell}>{name}</span>
                      <span className={styles.productCell}>{price}$</span>
                      <span
                        className={`${styles.productCell} ${styles.alignCenter}`}
                      >
                        {quantity}
                      </span>
                    </li>
                  );
                })}
              </ol>
            </ul>
            <div className={styles.sum}>
              <span className={styles.sumCell}>Sum</span>
              <span className={styles.sumCell}>{productsPrice}$</span>
            </div>
          </div>
          <div className={styles.payment}>
            <div>
              <h3 className={styles.paymentTitle}>Payment</h3>
            </div>
            <PayPalScriptProvider options={initialOptions}>
              <div className={styles.paymentButtons}>
                <PayPalButtons
                
                  createOrder={(data, actions) => {
                    return actions.order.create({
                      purchase_units: [
                        {
                          amount: {
                            value: productsPrice,
                          },
                        },
                      ],
                    });
                  }}
                  onApprove={onApprove}
                />
              </div>
            </PayPalScriptProvider>
          </div>
        </div>
        <div className={styles.shipping}>
          <div>
            <h3 className={styles.shippingTitle}>Shopping address</h3>
          </div>
          {Object.values(shipping).map(({ title, value }) => (
            <div key={uniqid()} className={styles.wrapper}>
              <span>{title}:</span>
              <span>{value}</span>
            </div>
          ))}
        </div>
        <div className={styles.menu__item}>
          <div className={styles.marquee}>
            <div className={styles.marquee__inner} aria-hidden="true">
              <span>Discounts</span>
              <span>Discounts</span>
              <span>Discounts</span>
              <span>Discounts</span>
            </div>
          </div>
        </div>
      </div>

      <Footer />
    </div>
  );
};

const mapStateToProps = (state) => {
  const { products, shippingAddress, user } = state;
  return { products, shippingAddress, user };
};

const mapDispatchToProps = {
  clearCart: productActions.clearCart,
};

export default connect(mapStateToProps, mapDispatchToProps)(OrderPageTemplate);
