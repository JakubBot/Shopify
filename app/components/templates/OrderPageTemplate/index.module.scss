@import 'styles/app.scss';

.orderContainer {
  position: relative;
  display: flex;
  flex-direction: column;
  min-height: 100vh;
}

.order {
  flex: 1 1 auto;
  margin-bottom: 40px;
  padding: 0 36px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  flex-direction: column;
  overflow: hidden;

  // max-width: 400px;
  @include phone {
    justify-content: space-around;
    padding: 20px 86px;
  }
  @include lDesktop {
    padding: 30px 186px;
  }
}
.orderWrapper {
  @include phone {
    margin-top: 10px;
    display: grid;
    grid-template-columns: 2fr 2fr;
  }

  @include sDesktop {
    width: 100%;
    grid-template-columns: 0.8fr 1.5fr;
  }
}
.payment {
  align-self: flex-start;
  z-index: 10;
  justify-self: center;
  margin-top: 35px;

  @include tablet {
    padding: 0 15px 0;
    margin-top: 0;
  }
}
.paymentButtons {
  max-width: 350px;
  margin: 30px 10px 0;
  overflow-y: auto;
  max-height: 300px;

  @include phone {
    max-height: unset;
    height: 300px;
  }
}

.orderProducts {
  max-width: 450px;
}
.orderProducts,
.payment,
.shipping {
  width: 100%;
}
.orderTitle {
  font-size: clamp(26px, 3vw, 44px);
  color: rgb(17, 17, 21);
  margin-bottom: 10px;

  @include tablet {
    margin-bottom: 30px;
  }
}
.productList {
  margin: 0;
  padding: 0;
  max-height: 220px;
  overflow-y: scroll;
  @include phone {
    max-height: unset;
    overflow-y: visible;
  }
}
.product {
  list-style-type: none;
  width: 100%;
  padding-right: 30px;
  display: grid;
  grid-template-columns: 2.4fr 1fr 1fr;
  margin-top: 20px;
}

.productCell {
  justify-self: flex-start;
  align-self: center;
  font-size: 16px;
}

.productName,
.productPrice,
.productQuantity {
  font-weight: 600;
  font-size: 17px;
  justify-self: flex-start;
}
.alignCenter {
  justify-self: center;
}

.paymentTitle {
  font-size: clamp(26px, 3vw, 44px);
  color: rgb(17, 17, 21);
}
.shipping {
  padding: 40px 0 45px;
  @include phone {
    padding: 0px 0 45px;
  }
}
.paymentButtons > div   {
  z-index: 1 !important;
  position: relative;
}

.shippingTitle {
  font-size: clamp(24px, 2.4vw, 38px);
  color: rgb(17, 17, 21);
  margin-bottom: 22px;
}

.wrapper {
  width: 100%;
  max-width: 450px;
  margin: 10px 0;
  display: grid;
  grid-template-columns: 2fr 2fr 1fr;
}
.sum {
  font-weight: 600;
  width: 100%;
  display: grid;
  grid-template-columns: 1.5fr 1fr 2fr;
  margin: 22px 0 0;
}

.menu__item {
  display: none;
  position: relative;
  right: 0px;
  transform: rotate(-35deg);

  @include sDesktop {
    transform: rotate(-30deg);
    right: -50px;
    display: initial;
  }
}
.marquee {
  font-size: 60px;
  position: absolute;
  top: 0;
  left: 0;
  --offset: 20vw;
  --move-initial: calc(-25% + var(--offset));
  --move-final: calc(-50% + var(--offset));
  font-style: italic;
}
.marquee__inner {
  width: fit-content;
  display: flex;
  position: relative;
  transform: translate3d(var(--move-initial), 0, 0);
  animation: marquee 5s linear infinite;
  animation-play-state: paused;
  opacity: 0;
  transition: opacity 0.1s;
}
.marquee span {
  padding: 0 2vw;
  -webkit-text-stroke: 1.5px rgb(0, 0, 0);
  -webkit-text-fill-color: transparent;
}
.menu__item__link {
  transition: opacity 0.4s;
  -webkit-text-stroke: 1.5px rgb(0, 0, 0);
  -webkit-text-fill-color: transparent;
  position: relative;
  z-index: 2;
}

.marquee__inner {
  transition-duration: 0.4s;
  animation-play-state: running;
  opacity: 1;
}

@keyframes marquee {
  0% {
    transform: translate3d(var(--move-initial), 0, 0);
  }

  100% {
    transform: translate3d(var(--move-final), 0, 0);
  }
}
