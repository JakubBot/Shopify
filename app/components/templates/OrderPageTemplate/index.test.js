import React from 'react';
import OrderPageTemplate from './index';
import renderWithRedux from 'components/mocks/renderWithRedux';

let mockProduct = {
  name: 'Buckle Bunny',
  slug: 'buckle-bunny',
  image: '/products/buckleBunny.webp',
  price: 4.99,
  isFeatured: false,
  id: 1,
  quantity: 1
};

const initialState = {
  products: [mockProduct],
  user: {
    name: '',
    email: '',
    token: '',
  },
  shippingAddress: {
    fullName: '',
    address: '',
    city: '',
    postalCode: '',
  },
};

it('Should render a product', () => {
  const { getByLabelText, ...tree } = renderWithRedux(<OrderPageTemplate />, {
    initialState,
  });
  let list = getByLabelText(/products/i);

  expect(list).toContainHTML(
   `<ol aria-label="products" class="productList"><li class="product"><span class="productCell">${mockProduct.name}</span><span class="productCell">${mockProduct.price}$</span><span class="productCell alignCenter">${mockProduct.quantity}</span></li></ol>`
  );
  expect(tree).toMatchSnapshot();
});

  /*
   <ol aria-label="products" class="productList">
    <li key={mockProduct.id} class="product">
     <span class="productCell">{mockProduct.name}</span>
       <span class="productCell">{mockProduct.price}$</span>
     <span class='productCell alignCenter'>{mockProduct.quantity}</span>
     </li> 
   </ol>
*/
