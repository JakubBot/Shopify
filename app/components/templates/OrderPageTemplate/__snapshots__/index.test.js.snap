// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render a product 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="orderContainer"
      >
        <header
          class="header"
        >
          <a
            href="/"
          >
            <h2
              class="title"
            >
              Shop
              <span
                class="emptySpan"
              >
                ify
              </span>
            </h2>
          </a>
          <nav
            class="nav"
          >
            <ul
              class="list"
            >
              <li
                class="listItem"
                data-scroll="slider"
              >
                Our products
              </li>
              <li
                class="listItem"
                data-scroll="about"
              >
                About us
              </li>
              <li
                class="listItem"
                data-scroll="contact"
              >
                Contact
              </li>
              <li
                class="listItem"
              >
                <a
                  href="/login"
                >
                   Login
                </a>
              </li>
            </ul>
            <div
              class="circle"
            />
          </nav>
        </header>
        <div
          class="order"
        >
          <div
            class="orderWrapper"
          >
            <div
              class="orderProducts"
            >
              <h3
                class="orderTitle"
              >
                Your Products
              </h3>
              <ul
                class="productList"
              >
                <li
                  class="product"
                >
                  <span
                    class="productCell productName"
                  >
                    Name
                  </span>
                  <span
                    class="productCell productPrice"
                  >
                    Price
                  </span>
                  <span
                    class="productCell productQuantity"
                  >
                    Quantity
                  </span>
                </li>
                <ol
                  aria-label="products"
                  class="productList"
                >
                  <li
                    class="product"
                  >
                    <span
                      class="productCell"
                    >
                      Buckle Bunny
                    </span>
                    <span
                      class="productCell"
                    >
                      4.99
                      $
                    </span>
                    <span
                      class="productCell alignCenter"
                    >
                      1
                    </span>
                  </li>
                </ol>
              </ul>
              <div
                class="sum"
              >
                <span
                  class="sumCell"
                >
                  Sum
                </span>
                <span
                  class="sumCell"
                >
                  4.99
                  $
                </span>
              </div>
            </div>
            <div
              class="payment"
            >
              <div>
                <h3
                  class="paymentTitle"
                >
                  Payment
                </h3>
              </div>
              <div
                class="paymentButtons"
              >
                <div
                  class=""
                />
              </div>
            </div>
          </div>
          <div
            class="shipping"
          >
            <div>
              <h3
                class="shippingTitle"
              >
                Shopping address
              </h3>
            </div>
            <div
              class="wrapper"
            >
              <span>
                Full Name
                :
              </span>
              <span />
            </div>
            <div
              class="wrapper"
            >
              <span>
                Address
                :
              </span>
              <span />
            </div>
            <div
              class="wrapper"
            >
              <span>
                City
                :
              </span>
              <span />
            </div>
            <div
              class="wrapper"
            >
              <span>
                Postal Code
                :
              </span>
              <span />
            </div>
          </div>
          <div
            class="menu__item"
          >
            <div
              class="marquee"
            >
              <div
                aria-hidden="true"
                class="marquee__inner"
              >
                <span>
                  Discounts
                </span>
                <span>
                  Discounts
                </span>
                <span>
                  Discounts
                </span>
                <span>
                  Discounts
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="footer"
        >
          <div
            class="footerLinks"
          >
            <h3
              class="footerLink"
            >
              <a
                href="/"
              >
                Promotions
              </a>
            </h3>
            <h3
              class="footerLink"
            >
              <a
                href="/"
              >
                Employees
              </a>
            </h3>
            <h3
              class="footerLink"
            >
              <a
                href="/"
              >
                Terms
              </a>
            </h3>
          </div>
          <div
            class="footerPhone"
          >
            <svg
              class="phone"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1.885.511a1.745 1.745 0 0 1 2.61.163L6.29 2.98c.329.423.445.974.315 1.494l-.547 2.19a.678.678 0 0 0 .178.643l2.457 2.457a.678.678 0 0 0 .644.178l2.189-.547a1.745 1.745 0 0 1 1.494.315l2.306 1.794c.829.645.905 1.87.163 2.611l-1.034 1.034c-.74.74-1.846 1.065-2.877.702a18.634 18.634 0 0 1-7.01-4.42 18.634 18.634 0 0 1-4.42-7.009c-.362-1.03-.037-2.137.703-2.877L1.885.511z"
                fill-rule="evenodd"
              />
            </svg>
            <h3>
              772567123
            </h3>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="orderContainer"
    >
      <header
        class="header"
      >
        <a
          href="/"
        >
          <h2
            class="title"
          >
            Shop
            <span
              class="emptySpan"
            >
              ify
            </span>
          </h2>
        </a>
        <nav
          class="nav"
        >
          <ul
            class="list"
          >
            <li
              class="listItem"
              data-scroll="slider"
            >
              Our products
            </li>
            <li
              class="listItem"
              data-scroll="about"
            >
              About us
            </li>
            <li
              class="listItem"
              data-scroll="contact"
            >
              Contact
            </li>
            <li
              class="listItem"
            >
              <a
                href="/login"
              >
                 Login
              </a>
            </li>
          </ul>
          <div
            class="circle"
          />
        </nav>
      </header>
      <div
        class="order"
      >
        <div
          class="orderWrapper"
        >
          <div
            class="orderProducts"
          >
            <h3
              class="orderTitle"
            >
              Your Products
            </h3>
            <ul
              class="productList"
            >
              <li
                class="product"
              >
                <span
                  class="productCell productName"
                >
                  Name
                </span>
                <span
                  class="productCell productPrice"
                >
                  Price
                </span>
                <span
                  class="productCell productQuantity"
                >
                  Quantity
                </span>
              </li>
              <ol
                aria-label="products"
                class="productList"
              >
                <li
                  class="product"
                >
                  <span
                    class="productCell"
                  >
                    Buckle Bunny
                  </span>
                  <span
                    class="productCell"
                  >
                    4.99
                    $
                  </span>
                  <span
                    class="productCell alignCenter"
                  >
                    1
                  </span>
                </li>
              </ol>
            </ul>
            <div
              class="sum"
            >
              <span
                class="sumCell"
              >
                Sum
              </span>
              <span
                class="sumCell"
              >
                4.99
                $
              </span>
            </div>
          </div>
          <div
            class="payment"
          >
            <div>
              <h3
                class="paymentTitle"
              >
                Payment
              </h3>
            </div>
            <div
              class="paymentButtons"
            >
              <div
                class=""
              />
            </div>
          </div>
        </div>
        <div
          class="shipping"
        >
          <div>
            <h3
              class="shippingTitle"
            >
              Shopping address
            </h3>
          </div>
          <div
            class="wrapper"
          >
            <span>
              Full Name
              :
            </span>
            <span />
          </div>
          <div
            class="wrapper"
          >
            <span>
              Address
              :
            </span>
            <span />
          </div>
          <div
            class="wrapper"
          >
            <span>
              City
              :
            </span>
            <span />
          </div>
          <div
            class="wrapper"
          >
            <span>
              Postal Code
              :
            </span>
            <span />
          </div>
        </div>
        <div
          class="menu__item"
        >
          <div
            class="marquee"
          >
            <div
              aria-hidden="true"
              class="marquee__inner"
            >
              <span>
                Discounts
              </span>
              <span>
                Discounts
              </span>
              <span>
                Discounts
              </span>
              <span>
                Discounts
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="footerLinks"
        >
          <h3
            class="footerLink"
          >
            <a
              href="/"
            >
              Promotions
            </a>
          </h3>
          <h3
            class="footerLink"
          >
            <a
              href="/"
            >
              Employees
            </a>
          </h3>
          <h3
            class="footerLink"
          >
            <a
              href="/"
            >
              Terms
            </a>
          </h3>
        </div>
        <div
          class="footerPhone"
        >
          <svg
            class="phone"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1.885.511a1.745 1.745 0 0 1 2.61.163L6.29 2.98c.329.423.445.974.315 1.494l-.547 2.19a.678.678 0 0 0 .178.643l2.457 2.457a.678.678 0 0 0 .644.178l2.189-.547a1.745 1.745 0 0 1 1.494.315l2.306 1.794c.829.645.905 1.87.163 2.611l-1.034 1.034c-.74.74-1.846 1.065-2.877.702a18.634 18.634 0 0 1-7.01-4.42 18.634 18.634 0 0 1-4.42-7.009c-.362-1.03-.037-2.137.703-2.877L1.885.511z"
              fill-rule="evenodd"
            />
          </svg>
          <h3>
            772567123
          </h3>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
